proposals/0000-unlifted-datatypes.rst

```diff
@@ -0,0 +1,482 @@
+and ``Note [Implementation of UnliftedNewtypes]`` for details involving
+type-checking the parent data family.
+
+As usual for types of unlifted kind, values of unlifted data types may not be
+declared at the top-level. The top-level restriction is not fundamental (see
+`#17521 <https://gitlab.haskell.org/ghc/ghc/issues/17521>`_), but best
+discussed in a separate proposal. Note that we don't consider nullary
+constructors to be top-level bindings in that regard, so nullary unlifted
+constructors are allowed. There *is* an interesting interaction of this
+restriction with nullary data constructor *wrappers* (cf. "Effects and
+Interactions"), though, which we consider to be an implementation detail.
+
+Dynamic Semantics
+~~~~~~~~~~~~~~~~~
+
+Unliftedness (i.e., the absence of divergence) in general implies the need for
+an eager evaluation semantics, which GHC implements in expressions of kind
+``TYPE r``, where ``r`` is not ``BoxedRep Lifted`` (so
+``TYPE (BoxedRep Unlifted)`` in particular).
+
+Thus, call-by-value semantics are already well established within GHC. The
+novelty is pattern matching on and construction of unlifted data types, but
+that's exactly the same as it is for lifted data types. 
+
+We never have to decide on an evaluation strategy for ``TYPE (BoxedRep l)``
+with inconcrete ``l``, because the current restrictions on levity polymorphism
+don't allow to bind or pass expressions of this kind.
+
+Examples
+--------
+
+Unlifted (call-by-value) semantics
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Example:
+
+::
+
+ data UPair a b :: TYPE (BoxedRep Unlifted) where
+   UPair :: a -> b -> UPair a b
+
+* When occuring in a constructor field (e.g.
+  ``data T = MkT (UPair Int Bool)``), the semantics are identical to a strict
+  field (``data T = MkT !(Int, Bool)``).
+
+* In an application ::
+
+    f (if odd 42
+         then UPair a a
+         else UPair b b)
+
+  the unlifted argument is evaluated before before the application is beta
+  reduced. So call-by-value instead of call-by-need.
+
+* In a let binding ``let x = UPair a b in e`` ::
+
+    let x = if odd 42
+              then UPair a a
+              else UPair b b
+    in e
+
+  the unlifted right-hand side of ``x`` is evaluated before the body ``e``.
+
+The Strict data type
+~~~~~~~~~~~~~~~~~~~~
+
+We get to define ``Strict``
+
+::
+
+ data Strict a :: TYPE (BoxedRep Unlifted) where
+   Force :: !a -> Strict a
+
+that deprives itself and its argument of ⊥.
+
+``Strict`` is the very essence of this proposal: Every unlifted data type can
+be defined in terms of lifted data types and ``Strict``, at the cost of an
+additional indirection.
+
+Whether we could define ``Strict`` as a ``newtype`` -- which would then mean we
+coerce between kinds, meaning coercions aren't zero-cost -- is left for a
+separate proposal.
+
+Low-level code
```

Comment:
YOu give Haskell code, Core code, and C-- code for the bad version.  But, oddly, you do not give the Haskell code or Core code for the good version -- just the C-- code!   That makes it really hard to grok.

Reply:
Well, the diff in Core isn't particularly interesting (just substituting `Bool` for a new `SBool`). But I'll add it.